### PR TITLE
GPII-2666: Initial implementation of "gpii-diff"...

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-2666](https://issues.gpii.net/browse/GPII-2666) for details.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,43 +1,45 @@
...
-        "gpii-launcher": "1.0.0-dev.20170712T140305Z.aa60d4a",
-        "gpii-testem": "1.0.1",
-        "grunt": "1.0.1",
-        "grunt-markdownlint": "1.0.42",
-        "grunt-jsonlint": "1.1.0",
-        "istanbul": "0.4.5",
-        "node-jqunit": "1.1.7",
-        "rimraf": "2.6.2"
-    }
+  "name": "gpii-diff",
```
Comment (amb26) [inline package.json]: Can we have this back to 4-spacing?

File: src/js/isDiffArray-helper.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+*/
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.diff.helper.isDiffArray");
+
+gpii.diff.helper.isDiffArray.getHelperFunction = function () {
+    return function (diffDef, options) {
+        // TODO: Convert this to a single fluid.get check once https://issues.fluidproject.org/browse/FLUID-6217 is resolved.
```
Comment (amb26) [inline src/js/isDiffArray-helper.js]: Release 3.0.0-dev.20171102T191539Z.c0636b395 should have this resolved

File: src/js/diff.js [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+    }
+};
+
+/**
+ *
+ * Go through a string and create a report that can be used to render a "diff" view.  The result is an array of
+ * segments, which consist of a segment value and details about whether the segment is unchanged, added, or removed
+ * when compared to `leftString`.
+ *
+ * @param `leftString` - The first string, from whose perspective the "diff" results will be represented.
```
Comment (amb26) [inline src/js/diff.js]: Types for all parameters for these functions - {String}, {Array}, etc.

File: src/js/diff.js [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+ * @return {Object} - See below.
+ *
+ * The object returned has three keys:
+ *
+ * `leftIndex`:  The position of the matching segment in `leftArray`.  If there is no matching segment, this will be `-1`.
+ * `rightIndex`: The position of the matching segment in `rightArray`.  If there is no matching segment, this will be `-1`.
+ * `segment`:    The longest set of matching elements (in order) common to both arrays.  If there are multiple sets of the same length, the first is returned.
+ *
+ */
+gpii.diff.longestCommonArraySegment = function (leftArray, rightArray) {
```
Comment (amb26) [inline src/js/diff.js]: The comment should note some implementation subtleties, which should perhaps even be surfaced to top-level comments. Whilst exact equality throughout this algorithm is done on a deep equality basis, this algorithm uses only shallow equality as operated by "indexOf" as invoked on line 378. This means that situations like
leftArray: [{a: 1}, {b: 2}, 3]
rightArray: [{a: 1}, {b: 2}]
will be treated as a total diff. I don't propose fixing this right now but the user should be clear about the quality of support. Also it should be noted that this is a quadratic-time algorithm which may not be appropriate for very long arrays. Algorithms do exist which can compute this segment in O(n log n) time but again I am not proposing that we adopt one - but all the same the quality of implementation should be clearly flagged to the user.

File: src/js/diff.js [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+            segments.push({
+                value: trailingMatches[1],
+                index: originalString.length - trailingMatches[1].length
+            });
+        }
+
+        var combinedSegments = [];
+        for (var a = 0; a < segments.length; a++) {
+            for (var b = a + 1; b <= segments.length; b++) {
+                if (a + b > 0) {
```
Comment (amb26) [inline src/js/diff.js]: This is a very strange condition which looks like it is always true. Perhaps a typo or remnant of an earlier impl which had different loop bounds?

File: src/js/diff.js [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+    }
+    else {
+        return 1;
+    }
+};
+
+/**
+ *
+ * Break down a string into an array of "phrase segments", ordered from longest (the original string) to shortest (the
+ * shortest individual word).  Each "phrase segment" looks like:
```
Comment (amb26) [inline src/js/diff.js]: The reader should be warned that this is an algorithm which is quadratic in both time and space and should not be called with terribly long input.

File: src/js/diff.js [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+            }
+        });
+        if (firstMatchDef) {
+            matchDef = firstMatchDef;
+        }
+    }
+    return matchDef;
+};
+
+gpii.diff.sortPhraseDefsByLength = function (a, b) {
```
Comment (amb26) [inline src/js/diff.js]: This comparator does more than is advertised in the comments to "extractPhraseSegments" (it breaks ties based on index position) - this should be noted and explained

File: src/js/diff.js (hunk 4 above)
Comment (amb26) [reply src/js/diff.js]: After a little research: It should be noted that this is actually a *cubic* time algorithm (taking into account the costs of the arraysEqual call on line 381) - it could be reduced to quadratic time by implementing the standard LCS algorithm described on Wikipedia, or to n log n by some heroic efforts involving the "Method of Four Russians" ...

File: src/js/isDiffArray-helper.js (hunk 2 above)
Comment (the-t-in-rtf) [reply src/js/isDiffArray-helper.js]: Confirmed.  Thanks for the quick fix on that.

File: package.json (hunk 1 above)
Comment (the-t-in-rtf) [reply package.json]: Done.

File: src/js/diff.js [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+ * {
+ *  value: "a word or phrase, possibly including whitespace ",
+ *  index: 0 // The position of the phrase in `originalString`
+ * }
+ *
+ * @param originalString {String} - The original string to break down.
+ * @returns {Array} - An array of "phrase segments".
+ *
+ */
+gpii.diff.extractPhraseSegments = function (originalString) {
```
Comment (amb26) [inline src/js/diff.js]: There's enough code in your "extractPhraseSegments", "compareStrings", "compareArrays" and "longestCommonArraySegment" functions combined to at least consider the option of replacing them all with the standard LCS algorithm - on the plus side this would make this package suitable for differencing more substantial quantities of data and perhaps rationalise the implementation somewhat - on the minus side are the cognitive costs of maintaining, comprehending the algorithm and its various corner cases. I found an impl which has at least one test case (!) here: https://github.com/erictrinh/lcs/blob/master/index.js
I think if we don't, we should at least advertise that we are not using this standard algorithm in the various corresponding places in the README since I think it has become an expectation for someone using a utility named "diff" which operates on sequence-based data.

File: src/js/diff.js (hunk 7 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: I have renamed it and added JSDocs.

File: src/js/diff.js (hunk 5 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: Yes, previously it was possible for a and b to both be zero.  Removed.

File: src/js/diff.js (hunk 4 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: I have added that as a test case and will work on an approach that handles it better. 

As these "quadratic time" functions are called implicitly by various other functions, I will have to add a suitable warning throughout.

File: src/js/diff.js (hunk 3 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: Done.

File: src/js/diff.js (hunk 8 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: The approach described in the article and implemented in the linked repo unfortunately looks for the longest common subsequence irrespective of whether the members are adjacent.  So, when comparing the strings "abc" and "acd", that library returns "ac" as the longest common sub-segment.

One option would be to use an approach from [the related article on substrings](https://en.wikipedia.org/wiki/Longest_common_substring_problem, which I'll need to review for a  bit.

We might also (as I've done in previous approaches) break a string down into segments of blocks of word characters interleaved with blocks of non-word characters.  Once we have that, we can use the "standard" array comparison.

To use this, we'd need a little more than just the longest common sequence.  We need to know the position of each member of the sequence in each compared sequence, so that we can represent the removed/added material in context.

We also need to know how ties are broken, as there may be multiple differing longest sequences, as when comparing "xaybzc" and "abcxyz".  It's not 100% clear to me if the "standard" approach is "longest first" or not, or uses another mechanism to break ties.

Comment (amb26) [reply src/js/diff.js]: Yes, this is a important subtlety to note - it looks like we do indeed need two algorithms rather than one. The good news is that the substring problem is substantially easier than the subsequence problem and can be done in linear time.

Comment (the-t-in-rtf) [reply src/js/diff.js]: As we have discussed in the chat, I implemented an LCS approach and converted the string matching to use that.

**Comment by the-t-in-rtf** [comment]
I have made Markdown-It and Cheerio.js into optional dependencies.  @amb26, this should be ready for further review.

File: src/js/diff.js (hunk 6 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: It should be more reasonable now, I will confirm with the very long comparisons I need to make in the package that uses this to diff longer blocks of text.

File: src/js/diff.js (hunk 4 above)
Comment (the-t-in-rtf) [reply src/js/diff.js]: The new LCS approach specifically supports the case mentioned above.

**Comment by the-t-in-rtf** [comment]
Just to summarise recent discussions around this. We now are using a lookup table generator and traceback approach.  A full traceback to find all longest sequences does not scale well, it's possible to grind it to a halt with carefully crafted fairly small comparisons, such as comparing arrays like the following: 

```
var left =   [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25];
var right = [ 0, 0, 2, 2, 4, 4, 6, 6, 8, 8, 10, 10, 12, 12, 14, 14, 16, 16, 18, 18, 20, 20, 22, 22, 24, 24];
```

This highlights the full traceback's problem with multiple equivalent matches, i.e for every pair of "even" values on the right, the number of same-length sequences quadruples, as 0 and 2 might match index 1 and 3 of the right-hand array, or 1 and 4, or 2 and 3, or 2 and 4,

A traceback that returns any one of the sequences is orders of magnitude faster, as it only ever explores one path at a time.  However, the single path traceback still struggles once we are comparing thousands of array segments with multiple recurring values.  As we currently compare strings as interleaved blocks of word and non-word content, we end up exploding the number of potential matches just by virtue of attempt to match the spaces.  Although there are ways to work around this for the specific case of string comparisons, we cannot predict what material people might be comparing, particularly as we intend to use this when reporting failures in  `jqUnit.assertDeepEq` assertions.  Thus, Antranig and I have agreed that we need a safeguard to ensure that the comparison always eventually completes.

My current thinking is that a "full traceback" should be attempted, but timed out with a warning logged after 20 seconds.  If the full traceback times out, a "single path" traceback will be attempted, but timed out with a warning logged after another 20 seconds.

If both time out, an empty array will be returned.  This will result in the problematic change being represented as though all the original content was removed, and all the new content was added.

**Comment by amb26** [comment]
I think that probably 20 seconds is an excessive amount of time to commit to - especially bearing in mind that the typical usage of this algorithm will be as invoked during the render cycle of a GET request, and provide a clear vector for DoS attacks as well as significant user annoyance. In these terms we might think about devoting perhaps 500ms, or even less, to each phase. I'm quite sure that github, e.g. will set page render targets in units of 100ms rather than whole seconds ...

**Comment by amb26** [comment]
In parallel with this, we should also pursue the avenues left to us to shave a bit more off the overall computation time - e.g. the use of immutable data structures leading to "tail sharing" amongst the sequences, and avoiding the invocation of gpii.diff.equals after the table has been generated

**Comment by the-t-in-rtf** [comment]
@amb26, the timeout is now fully configurable.   I have been testing with 1000ms with good results, I agree that we should default low and allow people to relax the timeout as needed.

I also agree that we need to look at the use of gpii.diff.equals, as in adding instrumentation to explore the timing, I see that it often takes the bulk of the time involved in a "full" traceback.

**Comment by the-t-in-rtf** [comment]
Self-merging prior to moving to a new organisation.
